# PredDNAContam

PredDNAContam is a tool for DNA contamination prediction from biosample data.

## Input File Format (CSV)

When using **PredDNAContam**, your input data should be in CSV format with the following columns:

| Column  | Description |
|---------|------------|
| GQ      | Genotype quality |
| DP      | Total read depth |
| AF      | Allele frequency |
| VAF     | Variant allele frequency |

### Example CSV File
The CSV file is generated by extracting the following key features from a VCF (Variant Call Format) file for each variant.

Hereâ€™s an example of how the CSV should look after extracting these features:


```csv
GQ,DP,AF,VAF
20,47,0.5,0.23
60,25,0.5,0.24
23,55,0.5,0.78
```


### Example config.txt file:

Before running PredDNAContam, you need to configure the paths in the config.txt file. This file contains important directory paths and filenames, which should be set as follows: 

input_dir=/path/to/csv_files
output_dir=/path/to/output_directory/output_PredDNAcontam
model_filename=/path/to/PredDNAContam_model/Random_Forest_Contamination_Model.joblib
scaler_filename=/path/to/PredDNAContam_model_scaler/scaler.joblib


## Download and Installation

To install PredDNAContam, follow these steps:

1. Download the package
You can download the package from PyPI:
ðŸ‘‰ PredDNAContam on PyPI 
ðŸ”— https://pypi.org/project/PredDNAContam/#files

Download the file:
ðŸ“‚ preddnacontam-0.0.2.tar.gz 

2. Extract the package

```
After downloading, unzip the file:
tar -xvzf preddnacontam-0.0.2.tar.gz
cd preddnacontam-0.0.2
```

3. Install the package
Inside the extracted directory, run:

```
pip install .
```

## Running PredDNAContam
1. Modify the configuration file
Navigate to the scripts directory and update config.txt to set the correct paths for your input files, model, and output directory.

2. Run the tool 
After configuring the paths, execute PredDNAContam:

```
PredDNAContam
```

