# Target rules

targets = {'asan': [], 'pan': []}

if config['baczy']['args']['sequencing'] == 'paired':
    print ("Entering paired targets")
    targets['asan'].append(expand(os.path.join(dir_fastp_short,"{sample}_R1.fastq.gz"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_fastp_short,"{sample}_R2.fastq.gz"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_megahit, "{sample}-pr", "final.contigs.fa"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_megahit, "processed_assemblies", "{sample}_contigs.fa"), sample=sample_names))
    targets['asan'].append(os.path.join(dir_megahit, "checkm2", "quality_report.tsv"))
    targets['asan'].append(os.path.join(dir_megahit, "classify", "gtdbtk.bac120.summary.tsv"))
    targets['asan'].append(os.path.join(dir_megahit, "de_novo_output", "gtdbtk.bac120.decorated.tree"))
    targets['asan'].append(os.path.join(dir_megahit, "de_novo_output", "gtdbtk.bac120.tree.table"))
    targets['asan'].append(expand(os.path.join(dir_bakta_short, "{sample}_bakta", "{sample}.gbff"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_short, "{sample}_bakta", "{sample}.faa"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_short, "{sample}_bakta", "{sample}.fna"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_short, "{sample}_bakta", "{sample}.gff3"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_short, "{sample}_bakta", "{sample}.txt"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_short, "{sample}_bakta", "{sample}_amrfinderplus"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_short, "{sample}_prophages", "{sample}_prophage_prophage_coordinates.tsv"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_short, "{sample}_prophages", "{sample}_prophage_{sample}.gbff"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_short, "{sample}_prophages", "{sample}_phispy.log"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_short, "{sample}_bakta", "{sample}_defense_finder_systems.tsv"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_short, "{sample}_capsule", "macsyfinder.log"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_summary_short, "{sample}", "DONE"),sample=sample_names))
    targets['asan'].append(os.path.join(dir_summary_short, "defensefinder_summary.tsv"))
    targets['asan'].append(os.path.join(dir_summary_short, "prophage_regions.tsv"))
    targets['asan'].append(os.path.join(dir_summary_short, "checkm2_quality_report.tsv"))
    targets['asan'].append(os.path.join(dir_summary_short, "gtdbtk.bac120.summary.tsv"))
    targets['asan'].append(os.path.join(dir_summary_short, "bakta_summary.tsv"))
    targets['asan'].append(os.path.join(dir_summary_short, "amrfinder_summary.tsv"))
elif config['baczy']['args']['sequencing'] == 'longread':
    print ("entering nanopore targets")
    targets['asan'].append(os.path.join(dir_hybracter, "hybracter.out", "FINAL_OUTPUT", "hybracter_summary.tsv"))
    targets['asan'].append(expand(os.path.join(dir_hybracter, "hybracter.out", "final_assemblies", "{sample}_final.fasta"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_hybracter, "hybracter.out", "final_chromosomes", "{sample}_chromosome.fasta"), sample=sample_names))
    targets['asan'].append(os.path.join(dir_hybracter, "checkm2", "quality_report.tsv"))
    targets['asan'].append(os.path.join(dir_hybracter, "classify", "gtdbtk.bac120.summary.tsv"))
    targets['asan'].append(os.path.join(dir_hybracter, "de_novo_output", "gtdbtk.bac120.decorated.tree"))
    targets['asan'].append(os.path.join(dir_hybracter, "de_novo_output", "gtdbtk.bac120.tree.table"))
    targets['asan'].append(expand(os.path.join(dir_bakta_long, "{sample}_bakta", "{sample}.gbff"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_long, "{sample}_bakta", "{sample}.faa"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_long, "{sample}_bakta", "{sample}.fna"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_long, "{sample}_bakta", "{sample}.gff3"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_long, "{sample}_bakta", "{sample}.txt"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_long, "{sample}_bakta", "{sample}_amrfinderplus"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_long, "{sample}_prophages", "{sample}_prophage_prophage_coordinates.tsv"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_long, "{sample}_prophages", "{sample}_prophage_{sample}.gbff"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_long, "{sample}_prophages", "{sample}_phispy.log"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_long, "{sample}_bakta", "{sample}_defense_finder_systems.tsv"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_bakta_long, "{sample}_capsule", "macsyfinder.log"), sample=sample_names))
    targets['asan'].append(expand(os.path.join(dir_summary_long, "{sample}", "DONE"),sample=sample_names))
    targets['asan'].append(os.path.join(dir_summary_long, "defensefinder_summary.tsv"))
    targets['asan'].append(os.path.join(dir_summary_long, "prophage_regions.tsv"))
    targets['asan'].append(os.path.join(dir_summary_long, "checkm2_quality_report.tsv"))
    targets['asan'].append(os.path.join(dir_summary_long, "gtdbtk.bac120.summary.tsv"))
    targets['asan'].append(os.path.join(dir_summary_long, "bakta_summary.tsv"))
    targets['asan'].append(os.path.join(dir_summary_long, "amrfinder_summary.tsv"))

if config['baczy']['args']['sequencing'] == 'paired':
    targets['pan'].append(os.path.join(dir_panaroo_short, "summary_statistics.txt"))
    targets['pan'].append(os.path.join(dir_panaroo_short, "gene_presence_absence.csv"))
elif config['baczy']['args']['sequencing'] == 'longread':
    targets['pan'].append(os.path.join(dir_panaroo_long, "summary_statistics.txt"))
    targets['pan'].append(os.path.join(dir_panaroo_long, "gene_presence_absence.csv"))
